request-manager/spec/check-factory.go

```diff
@@ -0,0 +1,118 @@
+//
+// Warnings identify probable mistakes in the specs. For example, an arg
+// that is renamed twice using sets/arg/as syntax.
+// Warnings will be logged, but will not cause the RM to fail to boot.
+type CheckFactory interface {
+	MakeSequenceErrorChecks() ([]SequenceCheck, error)
+	MakeSequenceWarningChecks() ([]SequenceCheck, error)
+	MakeNodeErrorChecks() ([]NodeCheck, error)
+	MakeNodeWarningChecks() ([]NodeCheck, error)
+}
+
+// The absolute minimum of checks for the RM to work properly.
+type BaseCheckFactory struct {
+	AllSpecs Specs // All specs in specs dir
+}
+
+func (c BaseCheckFactory) MakeSequenceErrorChecks() ([]SequenceCheck, error) {
+	return []SequenceCheck{
+		RequiredArgsNamedSequenceCheck{},
+		OptionalArgsNamedSequenceCheck{},
+		StaticArgsNamedSequenceCheck{},
+
+		OptionalArgsHaveDefaultsSequenceCheck{},
+		StaticArgsHaveDefaultsSequenceCheck{},
+
+		ACLAdminXorOpsSequenceCheck{},
+		ACLsHaveRolesSequenceCheck{},
+		NoDuplicateACLRolesSequenceCheck{},
+	}, nil
+}
+
+func (c BaseCheckFactory) MakeSequenceWarningChecks() ([]SequenceCheck, error) {
+	return []SequenceCheck{}, nil
+}
+
+func (c BaseCheckFactory) MakeNodeErrorChecks() ([]NodeCheck, error) {
+	return []NodeCheck{
+		HasCategoryNodeCheck{},
+		ValidCategoryNodeCheck{},
+
+		SubsequencesExistNodeCheck{c.AllSpecs},
+
+		ValidEachNodeCheck{},
+		ArgsAreNamedNodeCheck{},
+		SetsAreNamedNodeCheck{},
+
+		ValidParallelNodeCheck{},
+
+		ConditionalHasIfNodeCheck{},
+		ConditionalHasEqNodeCheck{},
+		NonconditionalHasTypeNodeCheck{},
+
+		ValidRetryWaitNodeCheck{},
+
+		RequiredArgsProvidedNodeCheck{c.AllSpecs},
+	}, nil
+}
+
+func (c BaseCheckFactory) MakeNodeWarningChecks() ([]NodeCheck, error) {
+	return []NodeCheck{}, nil
+}
+
+// Some default checks. Not strictly necessary for proper operation of RM, but generally reasonable.
+type DefaultCheckFactory struct {
+	AllSpecs Specs // All specs in specs dir
+}
+
+func (c DefaultCheckFactory) MakeSequenceErrorChecks() ([]SequenceCheck, error) {
+	return []SequenceCheck{
+		NoDuplicateArgsSequenceCheck{},
+		RequiredArgsHaveNoDefaultsSequenceCheck{},
+		HasNodesSequenceCheck{},
+		NodesSetsUniqueSequenceCheck{},
+	}, nil
+}
+
+func (c DefaultCheckFactory) MakeSequenceWarningChecks() ([]SequenceCheck, error) {
+	return []SequenceCheck{}, nil
+}
+
+func (c DefaultCheckFactory) MakeNodeErrorChecks() ([]NodeCheck, error) {
+	return []NodeCheck{
+		ArgsExpectedUniqueNodeCheck{},
+		EachElementDoesNotDuplicateArgsExpectedNodeCheck{},
+		SetsAsUniqueNodeCheck{},
+
+		EachIfParallelNodeCheck{},
+
+		ConditionalNoTypeNodeCheck{},
+		NonconditionalNoIfNodeCheck{},
+		NonconditionalNoEqNodeCheck{},
+
+		RetryIfRetryWaitNodeCheck{},
+
+		NoExtraSequenceArgsProvidedNodeCheck{c.AllSpecs},
+	}, nil
+}
+
+func (c DefaultCheckFactory) MakeNodeWarningChecks() ([]NodeCheck, error) {
+	return []NodeCheck{
+		EachNotRenamedTwiceNodeCheck{},
+		ArgsNotRenamedTwiceNodeCheck{},
+		EachElementDoesNotDuplicateArgsExpectedNodeCheck{},
```

Comment:
This was supposed to be an error--looks like I had a had a copy/paste error and added that instead of `EachListDoesNotDuplicateArgsGivenNodeCheck` as a warning.

In that case, it just gets overwritten [here](https://github.com/square/spincycle/blob/spinc-linter/request-manager/chain/creator.go#L235-L242). I'm inclined to keep it as a default check, because it won't actually break the RM. Base checks currently enforce assumptions in template/chain, and if any assumption is false, template/chain segfaults.